Using ConcurrentHashMap for synchronizing LTI user login

otti-ssystems opened the pull request:
This Pull Request is correcting 

https://github.com/opencast/opencast/pull/1532 
and 
https://github.com/opencast/opencast/pull/1622

Now we simple using a ConcurrentHashMap to synchronize the user authenticated by LTI.
We save the username in the map and removing it in the finally block after user is added or updated.
The fix is related to https://github.com/opencast/opencast/pull/1532 and the mentioned exceptions.

The code of the patches above is not needed anymore.

otti-ssystems committed:
refs #4904 FIX using ConcurrentHashMap for LTI login User

lkiesow commented on modules/security-lti/src/main/java/org/opencastproject/security/lti/LtiLaunchAuthenticationHandler.java:
```diff
@@ -315,37 +317,36 @@ public Authentication createAuthentication(HttpServletRequest request, ConsumerA
-
-      Date loginDate = new Date();
+      if (attempts.putIfAbsent(username, Boolean.TRUE) != null) {
+        logger.warn("Concurrent access of user {}. Igoring.", username);
```
This now becomes expected behavior and is not something we need to warn users about, is it?

```suggestion
        logger.debug("Concurrent access of user {}. Igoring.", username);
```

lkiesow commented on modules/security-lti/src/main/java/org/opencastproject/security/lti/LtiLaunchAuthenticationHandler.java:
```diff
@@ -137,6 +137,9 @@
   private Set<String> usernameBlacklist = new HashSet<>();
 
+  /** concurrent attemtps */
+  private Map<String, Boolean> attempts = new ConcurrentHashMap<>(128);
```
Purely nitpicking but it's not that we count attempts so the choice for the name seems a bit odd.
;aybe:

```suggestion
  private Map<String, Boolean> activePersistence = new ConcurrentHashMap<>(128);
```

lkiesow reviewed (commented):
I did not try this so far but the mechanism seems valid and should work well.
Can you fix the build problem and maybe take a look at the two suggestions?

lkiesow committed:
Fix LTI build and naming

- Fix build
- warn to debug logging
- Fix typo
- Improve name of HashMap

lkiesow commented:
Hopefully, this should now build properly.

lkiesow reviewed (approved):
Tested this and works as expected. @wsmirnow or @gregorydlogan, do you want to test this as well? For me, it's fine like this and can be merged at any time.

lkiesow committed:
Concurrent LTI Requests

This patch adds an additional safeguard for handling concurrent
authentication requests executed on multiple server.

wsmirnow reviewed (approved):
It's fine for me.

gregorydlogan commented:
I haven't had a chance to test, but it looks good to me too.

lkiesow commented:
Okay, everyone approves 🎉
Then, if no one complains until EOD tomorrow I'll merge this.

lkiesow commented:
No one else complains → merging this
